presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,129 @@
+
+import static com.facebook.presto.SystemSessionProperties.getJoinDistributionType;
+import static com.facebook.presto.sql.planner.iterative.rule.ReorderJoinsUtils.createJoinAccordingToPartitioning;
+import static com.facebook.presto.sql.planner.iterative.rule.ReorderJoinsUtils.generatePartitions;
+import static com.facebook.presto.sql.planner.iterative.rule.ReorderJoinsUtils.planPriorityQueue;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.google.common.base.Preconditions.checkState;
+
+public class ReorderJoins
+        implements Rule
+{
+    private final CostComparator costComparator;
+
+    public ReorderJoins(CostComparator costComparator)
+    {
+        this.costComparator = costComparator;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!(node instanceof JoinGraphNode)) {
+            return Optional.empty();
+        }
+
+        JoinGraphNode joinGraph = (JoinGraphNode) node;
+        JoinNode joinNode = new JoinEnumerator().chooseJoinOrder(joinGraph, idAllocator, symbolAllocator, lookup, session);
+        return Optional.of(joinNode);
+    }
+
+    private class JoinEnumerator
+    {
+        private final Map<JoinGraphNode, JoinNode> memo = new HashMap<>();
+
+        private JoinNode chooseJoinOrder(JoinGraphNode joinGraph, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Lookup lookup, Session session)
+        {
+            if (memo.containsKey(joinGraph)) {
+                return memo.get(joinGraph);
+            }
+            else {
+                PriorityQueue<JoinNode> joinOrders = planPriorityQueue(symbolAllocator, lookup, session, costComparator);
+                // TODO: eliminate cross joins
+                for (Set<Integer> partitioning : generatePartitions(joinGraph.getSources().size())) {
+                    joinOrders.add(createJoinNodeTree(joinGraph, partitioning, idAllocator, symbolAllocator, lookup, session));
+                }
+                checkState(joinOrders.size() >= 1, "joinOrders cannot be empty");
+                JoinNode node = joinOrders.poll();
+                memo.put(joinGraph, node);
+                return node;
+            }
+        }
+
+        private JoinNode createJoinNodeTree(JoinGraphNode joinGraph, Set<Integer> partitioning, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Lookup lookup, Session session)
+        {
+            JoinNode placeholderJoin = createJoinAccordingToPartitioning(joinGraph, partitioning, idAllocator);
+            PlanNode left = lookup.resolve(placeholderJoin.getLeft());
+            PlanNode right = lookup.resolve(placeholderJoin.getRight());
```

Comment:
I think when you get to the leaf nodes of the tree you could end up with a groupIdNode as the source